Add PVL interpreter

mlsteele opened the pull request:
Add an interpreter for PVL to the client. The script used is hardcoded into the client as `hardcodedPVL`.

As of this PR:
- CheckProof works for all services (manually tested)
- There’s a feature flag `UsePVL` that's off.
- The `ProofErrors` returned are ok, but not exactly the same as before (will fix).
- Custom errors per instruction are not yet supported.
- There are (almost) no tests for this. They will come in a later PR _before_ the flag is enabled.
- Existing tests pass.

@maxtaco @oconnor663

maxtaco commented on go/libkb/pvl_helpers.go:
```diff
@@ -0,0 +1,151 @@
+
+func pvlStringToService(service string) (keybase1.ProofType, ProofError) {
+	switch service {
+	case "twitter":
```
in the past, I went pretty far out of the way to put everything about the Foo services into `proof_support_foo`.  Can we still achieve that?

mlsteele replied:
How's this? Now it reads from the protocol names.

maxtaco commented on go/libkb/pvl_hardcoded.go:
```diff
@@ -0,0 +1,117 @@
+	hardcodedPVL = *wrapper
+}
+
+// func init() {
```
what's this?

mlsteele replied:
gone.

maxtaco commented on go/libkb/pvl_hardcoded.go:
```diff
@@ -0,0 +1,99 @@
+
+var hardcodedPVLString = `
+{
+  "pvl_version": 1,
```
Also it would be good to have a sequential `revision` keyword.

maxtaco commented on go/libkb/pvl_helpers.go:
```diff
@@ -0,0 +1,113 @@
+	return string(b), nil
+}
+
+// Get the HTML contents of all elements in a selection, concatenated by a space.
```
Go-standard commenting begins the comment with the name of the function.

Also, can you provide a reason of why this is needed --- In my mind, it's in Twitter, to mash up the text and the URL inside the `<a>` tag, right?

mlsteele replied:
That's not the reason, the `element.Text()` method will extract that just fine. `.Text` will take a single element and extract the text of its subtree. This is for when a CSS selector returns more than one element. We don't currently have CSS selectors that actually return more than one element, but they could so it has to do something. I was thinking that one place you might want to match multiple elements is if you're looking for a sig and you don't know what element it's in. In that case, concat with space would still allow FindBase64 to work.

maxtaco commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+}
+
+// Checkproof verifies one proof by running the pvl on the provided proof information.
+func CheckProof(g GlobalContextLite, pvl *jsonw.Wrapper, service keybase1.ProofType, link RemoteProofChainLink, h SigHint) ProofError {
```
I think in general we need a lot more debug logging

maxtaco commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+		perr = pvlRunScript(g, script, newstate())
+		if perr == nil {
+			return nil
+		}
```
will want to log individual errors here (to the `Debug` channel)

maxtaco commented:
I made a first pass through the code, this is looking really good!

oconnor663 commented on go/libkb/pvl_hardcoded.go:
```diff
@@ -0,0 +1,99 @@
+    ],
+    "hackernews": [
+      {
+        "assert_regex_match": "/https://hacker-news\\.firebaseio\\.com/v0/user/%{username_service}/about.json/i"
```
Maybe worth highlighting with some kind of `_comment` field that this

oconnor663 commented on go/libkb/pvl_hardcoded.go:
```diff
@@ -0,0 +1,99 @@
+  "pvl_version": 1,
+  "services": {
+    "coinbase": [
+      { "assert_regex_match": "/^https://coinbase\\.com/%{username_service}/public-key$/i" },
```
Any reason not to just make everything case-sensitive/

oconnor663 commented on go/libkb/pvl_helpers.go:
```diff
@@ -0,0 +1,113 @@
+)
+
+func pvlServiceToString(service keybase1.ProofType) (string, ProofError) {
+	for name, stat := range keybase1.ProofTypeMap {
```
Do we need both directions of maps? :-D

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+	selectorIsIndex := err == nil
+	selectorKey, err := selector.GetString()
+	selectorIsKey := err == nil && !selectorIsIndex
+	allness, err := selector.AtKey("all").GetBool()
```
Should we assert that the value at `"all"` is `true` here?

oconnor663 replied:
Also, it doesn't look like this selector currently gets used? What's the intended use case?

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+// Most failures here log instead of returning an error. If an error occurs, ([], nil) will be returned.
+// This is because a selector may descend into many subtrees and fail in all but one.
+func pvlRunSelectorJSONInner(g GlobalContextLite, object *jsonw.Wrapper, selectors []*jsonw.Wrapper) ([]string, ProofError) {
+	if len(selectors) == 0 {
```
This is the "we've recursively reached the end of the selectors list and should now return an actual value" branch, right? Good place for a comment.

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+// This is because a selector may descend into many subtrees and fail in all but one.
+func pvlRunSelectorJSONInner(g GlobalContextLite, object *jsonw.Wrapper, selectors []*jsonw.Wrapper) ([]string, ProofError) {
+	if len(selectors) == 0 {
+		s, err := pvlJSONStringOrMarshal(object)
```
If I'm understanding this function, it'll `JSON.stringify` what's left unless it's a string, in which case it'll return it unquoted? This leaves me with the feeling that "an object that's not a string" is an unexpected case here. Would it be safer / clearer to just return an error if what we have isn't a string? The other obvious use case I can imagine for `selector_json` would be to pull out an integer, but it doesn't like we need to do that yet. If we want to support both in a forward looking sorta way, maybe `pvlJSONStringOrIntToString` or something like that?

oconnor663 commented on go/libkb/pvl_helpers.go:
```diff
@@ -0,0 +1,113 @@
+	}
+}
+
+func pvlJSONStringOrMarshal(object *jsonw.Wrapper) (string, error) {
```
Commented below where we're using this. Maybe we should be more aggressive about calling non-strings an error?

oconnor663 commented on go/libkb/pvl_helpers.go:
```diff
@@ -0,0 +1,113 @@
+	len := selection.Length()
+	results := make([]string, len)
+	errs := make([]error, len)
+	var wg sync.WaitGroup
```
Why do we need the WaitGroup? Is `selection.Each` asynchronous? It [doesn't look like it is](https://github.com/PuerkitoBio/goquery/blob/master/iteration.go#L8).

oconnor663 commented on go/libkb/pvl_helpers.go:
```diff
@@ -0,0 +1,113 @@
+func pvlSelectionContents(selection *goquery.Selection, useAttr bool, attr string) (string, error) {
+	len := selection.Length()
+	results := make([]string, len)
+	errs := make([]error, len)
```
Any particular reason to preallocate these slices instead of just using `append`?

oconnor663 commented on go/libkb/pvl_helpers.go:
```diff
@@ -0,0 +1,113 @@
+			return "", err
+		}
+	}
+	return strings.Join(results, " "), nil
```
Are there any places where we currently expect our selectors to find more than one element? That sounds like something we might want to be strict about, if we don't think we need it. Or maybe strict by default and then allow an explicit "yes I want to join all these please" flag?

Apart from just being-strict-for-the-sake-of-being-strict, which I'll admit to being a fan of, I have the following specific problem in mind:

- We define a CSS selector to pull your tweet out of a page that also contains other people's response tweets (which must not be confused with yours).
- Twitter changes something about their CSS classes, so that our selector now _also_ matches other people's tweets.
- Everything appears to keep working on our end, because the resulting text is strictly greater than what we were matching against before, but a difficult-to-notice vulnerability has silently opened up.

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+		// Enforce prooftype-dependent variables.
+		webish := (service == keybase1.ProofType_DNS || service == keybase1.ProofType_GENERIC_WEB_SITE)
+		if webish {
+			vars.UsernameService = ""
```
We might get less confusing errors if we use a sentinel here like `"<<<INVALID>>>"` or whatever, instead of the empty string. Otherwise somewhere down the line we're going to see `assertion failed, "foobar" != ""` and forget what that means.

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+		if perr == nil {
+			return nil
+		}
+		errs = append(errs, perr)
```
Logging here is probably important, since we're not going to be able to return all of these. Alternatively, we could define a compound error type that concatenates the error messages from all the failures together. (Reporting just the first error might be confusing in a situation where a proof was intended to match the second script, and failed because of a spelling error there or something. There the second error would be much more relevant, but we have no way of knowing that.)

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+
+func pvlStepInstruction(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	for name, step := range PvlSteps {
+		if pvlJSONHasKey(ins, name) {
```
Is this the place where we need to prohibit multiple instruction names in a single `ins` map?

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+		return state, perr
+	}
+	match := re.FindStringSubmatch(state.ActiveString)
+	if len(match) < 2 {
```
What's the significance of 2 here? This is the first match, followed by all of its subgroups, is that right? Are we asserting that it has at least one subgroup?

mlsteele replied:
Yep, added this comment ```// Assert that the match matched and has at least one subgroup.```

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+					"DNS script cannot transform url")
+			case modeknown:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Script cannot transform after fetch")
```
Can scripts do a fetch, make some transformations/asserts, then do a different fetch, and make yet more transformations/asserts? Any reason to restrict it to one?

I can imagine keeping track of the active page in your head could get confusing, but I'm wondering if we've thought about it. The model we've built for ourselves here does seem like it could accommodate this sort of thing.

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+					outerr = fmt.Errorf("Substitution argument %v out of range of match", i)
+				}
+			} else {
+				// Unrecognized variable, do no substitution.
```
Why not aggressively fail?

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+func pvlSubstitute(template string, vars PvlScriptVariables, match []string) (string, error) {
+	var outerr error
+	// Regex to find %{name} occurrences.
+	re := regexp.MustCompile("%{[\\w0-9]+}")
```
Looks like `\w` already includes the digits? https://golang.org/pkg/regexp/syntax/

oconnor663 commented on go/libkb/pvl_hardcoded.go:
```diff
@@ -0,0 +1,99 @@
+      { "assert_find_base64": "sig" }
+    ],
+    "rooter": [
+      { "assert_regex_match": "/^https?://[\\w:_\\-\\.]+/_/api/1\\.0/rooter/%{username_service}/.*$/i" },
```
I think `\w` already includes underscore in Go. Not sure if that's true in every other language though. (It's a Perl-compatibility thing?)

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+	target, err := ins.AtKey(PvlAssertFindBase64).GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not assert target %v", ins)
```
Let's include the why part here. `...because only 'sig' is supported.` or something.

oconnor663 commented on go/libkb/pvl_interp.go:
```diff
@@ -0,0 +1,825 @@
+	}
+	s := strings.Join(results, " ")
+
+	state.ActiveString = s
```
One final thought from me about `ActiveString`. Miles and I have already talked about it, nothing new here, just writing it out:

    { "selector_json": [0, "data", "children", 0, "data", "author"] },
    { "assert_regex_match": "/^%{username_service}$/i" },

These lines from the reddit PVL above are implicitly operating on `ActiveString`. An alternative to doing that would be to have every operation (both loads/selectors, and asserts) operate on a named string. The simplest version of this idea might be disappointingly verbose:

    { "selector_json": [0, "data", "children", 0, "data", "author"], "name": "author_name" },
    { "assert_regex_match": "/^%{username_service}$/i", "name": "author_name" },

But if we pair it with some other cleanup, it's possible the result could end up both easier to read and easier to validate.

    [ "selector_json",      "author_name", [0, "data", "children", 0, "data", "author"] ],
    [ "assert_regex_match", "author_name", "/^%{username_service}$/i" ],

Not sure how this would apply to `regex_capture`. Currently that one reads the active string (unlike `selector_*`, which read the full fetched document), and also writes to the active string. A different version that reads and writes to two named variables would definitely be more verbose, but possibly more flexible and less confusing? If we ever decided we wanted to have proofs that fetch *two* documents, and compare strings *between* them, some kind of naming will probably be necessary. (Also gives you slightly better error messages for free.)

Anyway, end of brain dump.

mlsteele commented:
Thanks for all the comments. Here's round 2, have at it!

Here's what's changed:
@maxtaco 
- Rebased over ProofContext and file moves. Now there's a `pvl` package.
- `proofs` depends on `pvl` which depends on `libkb`. Also `proofs` depends on `libkb`.
- Added `revision` keyword
- Added more debugging info (Log on every instruction run, log on every error, most logs contain exactly which part of the script it's about. See `debug.go:debugWithState` and friends.)

@oconnor663 
- Do we need both directions of maps.. We could. I don't think it's necessary.
- JSON selectors now asserts `all` is true.
- The intended use case of `all` is for e.g. looking in an array at all the elements because you're not sure which one contains the goodies.
- Added recursion termination comment to `pvlRunSelectorJSONInner`.
- Changed `pvlJSONStringOrMarshal` into `pvlJSONStringSimple` which is stricter.
- Bid farewell to the `WaitGroup`
- Made CSS selector require `multi: true` if they want to allow returning multiple elements _in the end_.
- For variable substitution, instead of using a sentinel like "<<<INVALID>>>", the variables that do not exist for some proof types are now an error to use in those proof types. See `pvlSubstitute`.
- Return a compound error if there ar emultiple (multiple scripts)... but it has to have an error code, so I just grabbed the one from the first error. Is that good?
- Prohibit multiple instruction names in one instruction. See `pvlStepInstruction`

Not addressed yet:
(There's lots of stuff going on with this change, could you bring these up again if you're still interested?)
- @oconnor663 Why not make all url matches case sensitive?
- This doesn't support multiple fetches in one script. That seems like power I don't expect we'll need. It seemed bothersome that a rogue pvl script would be able to fire a ton of requests.
- Your comment about restructuring things to have instructions be lists and named registers.

Coming in the future, not this PR:
- Tests for helpers, tests of real-ish pvl
- rename all the methods to get rid of `pvl` suffix, because this has its own package now :)

mlsteele commented:
Also I squashed, forced pushed, moved files, and just generally made your lives harder. And I hope to repay you with this command, which should show differences between the version you reviewed and this version _for one file_.
```kdiff3 <(git show miles/CORE-3526-pvl-oldpr:go/libkb/pvl_interp.go) <(git show miles/CORE-3526-pvl:go/pvl/interp.go)```

maxtaco commented on go/libkb/errors.go:
```diff
@@ -45,6 +45,9 @@ type ProofErrorImpl struct {
 
 func NewProofError(s keybase1.ProofStatus, d string, a ...interface{}) *ProofErrorImpl {
+	if len(a) == 0 {
+		return &ProofErrorImpl{s, d}
```
comment here?

maxtaco commented on go/pvl/hardcoded.go:
```diff
@@ -0,0 +1,104 @@
+  "revision": 1,
+  "services": {
+    "coinbase": [
+      { "assert_regex_match": "/^https://coinbase\\.com/%{username_service}/public-key$/i" },
```
so in the end, we're using `{ "function" : "arg" }` style objects.  What was the winning argument?

oconnor663 replied:
Yeah. Miles convinced me that lists don't work well when we want different functions to take different sets of named arguments.

maxtaco commented on go/pvl/debug.go:
```diff
@@ -0,0 +1,42 @@
+
+func debug(g libkb.ProofContext, format string, arg ...interface{}) {
+	s := fmt.Sprintf(format, arg...)
+	g.GetLog().Debug("PVL %v", s)
```
I love this tooling for debugging -- however, the problem is that all file/line output will just say `debug.go:26`. There is a mechanism to change the stack frame we're looking relative to, you might want to plumb that through.

maxtaco commented on go/pvl/interp.go:
```diff
@@ -0,0 +1,908 @@
+)
+
+const (
+	PvlAssertRegexMatch = "assert_regex_match"
```
make a class for `PvlCommandName`?

maxtaco commented:
I had a few comments, but overall this looks great!

mlsteele commented:
Alright, here it is again
- fixed the logger stack offset by using `CloneWithAddedDepth` which unfortunately meant adding `ProofContextExt`.
- made `PvlCommandName` a type
- added a little comment in the double string interpolation fix

As for deciding about {} vs [] I will take to @oconnor663 offline today.

maxtaco commented:
OK, 👍 for me, but need one from @oconnor663 too

oconnor663 commented on go/pvl/helpers.go:
```diff
@@ -0,0 +1,127 @@
+				errs[i] = fmt.Errorf("Could not get attr %v of element", attr)
+			}
+		} else {
+			results[i] = element.Text()
```
Is it possible that one element in a selection contains another element that's also in the selection? Could that make the same `Text()` appear twice in our return value here, even though it only appears once in the actual HTML document? Do we care about that?

mlsteele replied:
Oh, I think that could be a possibility. To get into that situation you would have already explicitly added `multi: true` to the css instruction. This could definitely be fixed using goquery's `Contains` method to collapse the containment relations, but I don't think that's worth it.

oconnor663 commented on go/pvl/interp.go:
```diff
@@ -0,0 +1,911 @@
+	opts := template[lastSlash+1:]
+	if !regexp.MustCompile("[imsU]*").MatchString(opts) {
+		return nil, perr
+	}
```
Are there any cases where we want to guard against the same letter appearing twice? Or is that just kind of redundant but ultimately harmless?

oconnor663 commented on go/pvl/interp.go:
```diff
@@ -0,0 +1,911 @@
+
+	var outerr libkb.ProofError
+	// Regex to find %{name} occurrences.
+	re := regexp.MustCompile("%{[\\w]+}")
```
I think `{` and `}` are special characters in regex. Should we escape them?

oconnor663 commented on go/pvl/interp_test.go:
```diff
@@ -0,0 +1,84 @@
+	{false, keybase1.ProofType_TWITTER, nil,
+		"%{hostname}", ""},
+	{false, keybase1.ProofType_DNS, nil,
+		"%{username_service}", ""},
```
Can we test what happens if you do something stupid like `%{%{username_keybase}}` and your `username_keybase` is `hostname` or something else confusingly valid.

oconnor663 commented:
Just a few hypothetical questions above.

SHIP IT 🐑 🐑 🐑

mlsteele commented:
Rebased, answered a question, fixed the other things, waiting for tests :)
